File: src/state_handlers/initial.js
/**
 * @fileOverview
 * @author Peter Mitrano- mitrnanopeter@gmail.com
 */

'use strict';

var Core = require('../core');
var Alexa = require('../alexa');
var attributesHelper = require('../DynamoAttributesHelper');

/**
 * Represents the reponses when we aren't really in any particular state.
 * This is also the state we should begin in for every new session
 */
module.exports = Alexa.CreateStateHandler(Core.states.INITIAL_STATE, {
  /** User can start off by immediately asking for a recipe */
  'StartNewRecipeIntent': function () {
    this.emit("StartNewRecipeIntent" + Core.states.NEW_RECIPE);
  },
  /** The user says something like "Open my cookbook" */
  'LaunchRequest': function () {
    let ftu = Core.firstTimeIntroductionIfNeeded(this);
    if (!ftu) {
      if (this.event.session.new) {
        this.attributes.invocations += 1;
        this.handler.state = Core.states.NEW_RECIPE;
        this.emit(":ask", "Hi again. What do you want to make?");
        this.emit(':saveState', true);
        return true;
      }
      else {
        this.emit(":tell", "I've already been launched");
      }
    }
  },
  'SessionEndedRequest': function() {
    this.emit(":tell", "Goodbye!");
  },
  /** Any intents not handled above go here */
  'Unhandled': function() {
    if (this.event.session.new) {
      let ftu = Core.firstTimeIntroductionIfNeeded(this);
      if (!ftu) {
        this.emit(":tell", "I'm not sure what you want. Try asking to make something.");
      }
    }
    else {
      this.emit(":tell", "I'm not sure what you want. start by asking to make something.");
    }
  }
});

File: my_cookbook/skill/handlers/initial.py
import logging
from my_cookbook.util import core
from my_cookbook.util import responder


class InitialHandler():
    def SessionEndedRequest(self, handlers, persistant_attributes, attributes, slots):
        persistant_attributes[core.STATE_KEY] = core.States.INITIAL_STATE
        return responder.tell("Goodbye.")

    def LaunchRequest(self, handlers, persistant_attributes, attributes, slots):
        if persistant_attributes['invocations'] == 1:  # first time! Say hello!
            attributes[core.STATE_KEY] = core.States.ASK_TUTORIAL
            return responder.ask(
                "Hi, I'm your new cookbook. Would you like start off with a tutorial?", None,
                attributes)
        else:
            if attributes['new']:
                attributes[core.STATE_KEY] = core.States.NEW_RECIPE
                return responder.ask("Welcome back. What would you like to make?", None, attributes)
            else:
                persistant_attributes[core.STATE_KEY] = core.States.INITIAL_STATE
                return responder.tell("I've already been launched.")

    def Unhandled(self, handlers, persistant_attributes, attributes, slots):
        persistant_attributes[core.STATE_KEY] = core.States.INITIAL_STATE
        return responder.tell("We've already been talking" \
            " but I have no idea what about, so I will exit this session. Please" \
            " start over by saying, Alexa launch my cookbok.")


handler = InitialHandler()
state = core.States.INITIAL_STATE
